### PR TITLE
Heavily improve automatic model card generation + Patch XLM-R

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

## Pull Request overview
* Heavily improve automatic model card generation

## Example generated output
These model cards were fully generated - none of this was edited except for one sentence: `See [train.py](train.py) for the training script.`:
* [tomaarsen/span-marker_bert-base-uncased-keyphrase-inspec](https://huggingface.co/tomaarsen/span-marker_bert-base-uncased-keyphrase-inspec): no evaluation during training
* [tomaarsen/span-marker-bert-base-uncased-bionlp](https://huggingface.co/tomaarsen/span-marker-bert-base-uncased-bionlp): with evaluation during training

## Details
Add the following sections to the README body:
* Model Description (encoder model, maximum sequence length, maximum entity length, training dataset, language, license)
* Model Sources (repo link, thesis link)
* Model labels (a table)
* Direct uses (inference, a code snippet)
* Downstream use (finetuning, a code snippet)
* Training set metrics (sentence length, entities per sentence)
* Training hyperparameters
* Training results
* Environmental Impact (emissions, training time)
* Training hardware
* Framework versions

Beyond that, I added some values to the README metadata:
* language
* license
* library_name
* tags
* datasets
* metrics
* widget
* pipeline_tag
* co2_eq_emissions
* model-index (f1, recall, precision)
* base_model

The values for these are automatically computed via the Trainer and a `ModelCardCallback`.

### TODO
- [ ] Tests
- [ ] Additional documentation

cc: @davanstrien I think this might really interest you. Perhaps you have some useful feedback!

- Tom Aarsen

### CODE REVIEW

**Comment by stefan-it** [comment]
Hey  @tomaarsen ,

would be awesome if the `base_model` metadata tag could also be written, see discussion here https://twitter.com/vanstriendaniel/status/1699102884246479214 :hugs:

**Comment by tomaarsen** [comment]
That sounds like a great idea! 

@davanstrien I noticed that the Model Card metadata documentation is slightly lacking: https://huggingface.co/docs/hub/model-cards. In particular, e.g. `base_model` and `pipeline_tag` (and perhaps others) are missing. Notably, this [modelcard.md specification](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1) that the documentation links to seems incomplete (it also doesn't have `co2_eq_emissions`, etc.).

**Comment by davanstrien** [comment]
Thanks for the ping. This looks sick!! A few comments/questions/suggestions

- `train.py` to clarify, does this come from the spanmarker library? Will this also work if someone pushes from a notebook?
- Really happy to see how much metadata you are putting into the model cards automatically, super cool to see the model card attempting to find language info automatically! 

Suggestion: I love all the stuff that's captured in the model card at the moment. This is probably a matter of preference, but it might also be nice to include sections that would be nice for the user to complete i.e. bias. I think it's difficult to reliably do this automatically, but it could be nice to include this in the template to give people a gentle nudge to complete this. One way of doing this if you don't want to model card to get full of empty sections would be to include these parts wrapped in comments markers that way, people are encouraged to add this information when they are editing the card, but it doesn't clutter the visible card too much. 

>@davanstrien I noticed that the Model Card metadata documentation is slightly lacking: [huggingface.co/docs/hub/model-cards](https://huggingface.co/docs/hub/model-cards). In particular, e.g. base_model and pipeline_tag (and perhaps others) are missing. Notably, this [modelcard.md specification](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1&rgh-link-date=2023-09-06T06%3A27%3A31Z) that the documentation links to seems incomplete (it also doesn't have co2_eq_emissions, etc.).

I'll take a look at that today, IRC there are some parts of the model card which are validated hub side which might be why some of these aren't included but I'll confirm and get back to you. The base_model wasn't used regularly before but I'll make sure that gets added to the docs since I think this is a really useful piece of metadata and often very easy to generate for libraries focused on fine-tuning :)

**Comment by tomaarsen** [comment]
Hello!

The `train.py` I uploaded manually. I don't want to automatically include the "current file" into the HF model repository, as it may contain private keys or other data that should not be uploaded. I also included the sentence linking to `train.py` manually.

At the moment the language(s) are just user provided, but it would actually be interesting to see if the dataset metadata contains a language.

As for the commented out sections: that's a great idea. I'll definitely throw that on my TODO.

I appreciate your insights!

- Tom Aarsen

**Comment by davanstrien** [comment]
>At the moment the language(s) are just user provided, but it would actually be interesting to see if the dataset metadata contains a language. 

If it's a hub dataset, this would be fairly easy to do via the `huggingface_hub` library (you'd want to wrap this in some try/except blocks) 

```
from huggingface_hub import dataset_info
data_info = dataset_info('imdb')
data_info.cardData.get('language')
```
I personally think this would be nice to try and add this information automatically, but it might be unexpected for users and wouldn't necessarily work well for all datasets i.e. ones which have multiple languages which the user is only partially using (English split of a dataset for example). 

One potential option would be to surface this info to the user when they are pushing to the hub in a log message. Alternatively, this could also go into a 'suggested metadata' section that is commented out by default with an explanation of where the metadata comes from. i.e. something like 

```
---
tags:
- chemistry
library_name: generic
# From the dataset used to train this model (https://huggingface.co/datasets/imdb) 
# we have inferred the following languages as being likely for your model 
# If this is correct, uncomment the lines below. 
#language
# - en 
---
```

**Comment by tomaarsen** [comment]
Do you know if it's possible to have comments in the metadata section of a Hub README? 
I'll definitely try to look into the automated language info. Even if the automatically generated language is incomplete (e.g. it finds English but the model is trained for English, French and German), then that's already better than not having a language whatsoever.

- Tom Aarsen

**Comment by davanstrien** [comment]
> Do you know if it's possible to have comments in the metadata section of a Hub README? I'll definitely try to look into the automated language info. Even if the automatically generated language is incomplete (e.g. it finds English but the model is trained for English, French and German), then that's already better than not having a language whatsoever.

Yes (https://huggingface.co/davanstrien/test/raw/main/README.md), it's a bit ugly when you look at the raw version but the [rendered version](https://huggingface.co/davanstrien/test/blob/main/README.md) just doesn't display those parts.

**Comment by tomaarsen** [comment]
That repo is private, but I believe you on your word 😄

**Comment by davanstrien** [comment]
> That repo is private, but I believe you on your word 😄

sorry just made it public!

**Comment by tomaarsen** [comment]
That looks totally fine I think, I'll try to see if I can easily do something like that - I'm not sure as I currently use some yaml dumping utils for it.

**Comment by tomaarsen** [comment]
The language is now automatically inferred from the metadata from the dataset, when possible. 

Beyond that, I've added commented-out sections to the model card template in 179971f9703de0bb54fd8a1263c07b30866e2684. @davanstrien perhaps in the coming days you'll have some time to give your thoughts on the sections? They're inspired by [this docs page](https://huggingface.co/docs/hub/model-card-annotated), but they're not a 1 to 1 copy. 

I think this PR is starting to be feature complete - just the docs and tests remain.

- Tom Aarsen

**Comment by tomaarsen** [comment]
Note to self: Ensure that the model card is created correctly on `push_to_hub`.

File: span_marker/model_card_template.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+{%- endif %}
+{% if language -%}
+    - **Language{{"s" if language is not string and language | length > 1 else ""}}:**
+    {%- if language is string %} {{ language }}
+    {%- else %} {% for lang in language -%}
+            {{ lang }}{{ ", " if not loop.last else "" }}
+        {%- endfor %}
+    {%- endif %}
+{%- else -%}
+    <!-- - **Language:** Unknown -->
```
Comment (davanstrien) [inline span_marker/model_card_template.md:36]: This is very cool :)

File: span_marker/model_card_template.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+### Model Labels
+{{ label_examples }}{% endif -%}
+{% if metrics_table %}
+## Evaluation
+
+### Metrics
+{{ metrics_table }}{% endif %}
+## Uses
+
+### Direct Use
```
Comment (davanstrien) [inline span_marker/model_card_template.md:58]:
```suggestion
### Direct Use for Inference
```

Comment (davanstrien) [reply span_marker/model_card_template.md:58]: this is a small preference on my part, feel free to ignore :)

File: span_marker/model_card_template.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+)
+trainer.train()
+trainer.save_model("{{ model_id | default('span_marker_model_id', true) }}-finetuned")
+```
+</details>
+
+<!--
+### Out-of-Scope Use
+
+*List how the model may foreseeably be misused and address what users ought not do with the model.*
```
Comment (davanstrien) [inline span_marker/model_card_template.md:97]:
```suggestion
*List how the model may foreseeably be misused and address what users ought not to do with the model.*
```

File: span_marker/model_card_template.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+-->
+{% if tokenizer_warning %}
+### {{ warn_emoji }} Tokenizer Warning
+The [{{ encoder_name if encoder_name else encoder_id }}](https://huggingface.co/models/{{ encoder_id }}) tokenizer distinguishes between punctuation directly attached to a word and punctuation separated from a word by a space. For example, `Paris.` and `Paris .` are tokenized into different tokens. During training, this model is only exposed to the latter style, i.e. all words are separated by a space. Consequently, the model may perform worse when the inference text is in the former style.
+
+In short, it is recommended to preprocess your inference text such that all words and punctuation are separated by a space. Some potential approaches to convert regular text into this format are NLTK [`word_tokenize`](https://www.nltk.org/api/nltk.tokenize.word_tokenize.html) or spaCy [`Doc`](https://spacy.io/api/doc#iter) and join the resulting words with a space.
```
Comment (davanstrien) [inline span_marker/model_card_template.md:103]: Could be worth adding a small example to the SpanMarker docs page and linking to from here?

File: span_marker/model_card_template.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+{% endif %}
+<!--
+## Bias, Risks and Limitations
+
+*What are the known or foreseeable issues stemming from this model?*
```
Comment (davanstrien) [inline span_marker/model_card_template.md:108]:
```suggestion
*What are the known or foreseeable issues stemming from this model? You could also flag here known failure cases or weaknesses of the model.*
```

File: span_marker/model_card_template.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+{% endfor %}{% endif %}{% if eval_lines %}
+### Training Results
+{{ eval_lines }}{% endif %}{% if co2_eq_emissions %}
+### Environmental Impact
+Carbon emissions were measured using [CodeCarbon](https://github.com/mlco2/codecarbon).
+- **Carbon Emitted**: {{ "%.3f"|format(co2_eq_emissions["emissions"] / 1000) }} kg of CO2
+- **Hours Used**: {{ co2_eq_emissions["hours_used"] }} hours
+
+### Training Hardware
+- **On Cloud**: {{ "Yes" if co2_eq_emissions["on_cloud"] else "No" }}
```
Comment (davanstrien) [inline span_marker/model_card_template.md:132]: Awesome!

**Review by davanstrien — COMMENTED** [review]
This is looking really good.  I focused mainly on the model card template so didn't really review the code at all.

**Comment by tomaarsen** [comment]
> I focused mainly on the model card template so didn't really review the code at all.

That's no worries! I meant the template, so this is perfect :)

File: span_marker/model_card_template.md (hunk 1 above)
Comment (tomaarsen) [reply span_marker/model_card_template.md:36]: Haha, it took a while to get the Jinja to co-operate here.

File: span_marker/model_card_template.md (hunk 4 above)
Comment (tomaarsen) [reply span_marker/model_card_template.md:103]: Hmm, that is definitely worthwhile. I'll have to make a section for it in the docs regardless.

File: span_marker/model_card_template.md (hunk 6 above)
Comment (tomaarsen) [reply span_marker/model_card_template.md:132]: All from the `codecarbon` package, which is already integrated into `transformers`.
